### ISSUE TITLE
[$250] LOW: Investigate using native-stack on native platforms instead of stack

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#29948]
Coming from [here](https://expensify.slack.com/archives/C05LX9D6E07/p1697644880585569)

When we have refactored the App navigation, we have chose to use the normal react-navigation stack as that gave us more control over animation/ transitions of pages on web. This was to also not complicate the MVP for the project with two types of stack.

The problem is that the normal stack does not lead to native feel of app page transitions in native device which makes the feel of the app worse as the page transitions are ever so slightly off compared to the other native apps you might be using on daily basis.

Its important to address this discrepancy and ideally introduce the native-stack for native iOS and Android platforms, as chat/ P2P users have high standards for native app quality.

cc @hannojg @adamgrzybowski 

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021866177344381113052</li>
        <li>Upwork Job ID: 1866177344381113052</li>
        <li>Last Price Increase: 2024-12-09</li>
    </ul>
</details>

**Comment by hannojg** [comment]
Yes, will help investigating this or next week 🤝!

**Comment by hannojg** [comment]
PR is up: 
- https://github.com/Expensify/App/pull/29991

**Comment by hannojg** [comment]
We got two blockers at the minute:

- iOS 17 has bugs with the keyboard which hit in expensify once we turn on the native stack, waiting for an OS update here
- This upstream PR: https://github.com/software-mansion/react-native-screens/pull/1945 so we have the same native transitions on android (right now the native transitions on android feel outdated)

**Comment by WoLewicki** [comment]
@hannojg as for the second dot, the `ios like` do you think that this animation is what people want on `Android` to have the native feel? Maybe we could explore the other animations that are strictly taken from the `Android` XMLs?

Also, could you elaborate how does `native-stack` improve the performance of the app exactly? One thing I can think of is native swipe to go back on `iOS` which is controlled by the native code, but you still need to mount the previous screen on the native side when it happens, did you check how it performs when going back to reports with many many messages?

**Comment by hannojg** [comment]
> ios like do you think that this animation is what people want on Android to have the native feel

We have to double check the animation. Its called `iOS like` but I think it actually feels like the android default slide transition

> Also, could you elaborate how does native-stack improve the performance of the app exactly

This is more about perceived performance and UX feeling. Right now you can tell by using the expensify app, that the transitions aren't the native transitions. So we want to have the native transitions so expensify feels like a native app.

> you still need to mount the previous screen on the native side when it happens, did you check how it performs when going back to reports with many many messages?

Do I understand it correctly that you are suggesting that switching to native-stack could be a performance issue when doing the swipe back gesture?

**Comment by WoLewicki** [comment]
> Do I understand it correctly that you are suggesting that switching to native-stack could be a performance issue when doing the swipe back gesture?

We have the same abstraction in the `stack` navigator, where we control the mounting/unmounting the views by using `activityState` prop. Native-stack should not have worse performance, I'm just curious if it is fast enough no to be laggy in high traffic accounts 😅

**Comment by mountiny** [comment]
we should test it out with high traffic accounts, the E2E tests once are built in like 2 weeks should also tell us more how it performs there

**Comment by mountiny** [comment]
Keeping this monthly, its not high priority and still waiting on some issues to get resolved.

**Comment by mountiny** [comment]
@hannojg How is this looking now with latesst ios, have they fixed anything?

**Comment by hannojg** [comment]
Hey, I am handing this over to our Chief Keyboard Officer @kirillzyusko who is way more familiar with those keyboard issues. Once he is due with fixing [some other keyboard handling issues](https://github.com/Expensify/App/pull/16356#issuecomment-1896414974) he'll take a look!

**Comment by mountiny** [comment]
@kidroca Can you comment on this one?

**Comment by kirillzyusko** [comment]
I'm ready to pick up this task, so you can assign it on me 👀

**Comment by mountiny** [comment]
Sorry Kidroca, that was a typo!

@kirillzyusko Lets update the PR to see where we are at now

**Comment by Julesssss** [comment]
Assigning @abdulrahuman5196 based on [this comment](https://expensify.slack.com/archives/C02NK2DQWUX/p1708607683163729)

**Comment by kirillzyusko** [comment]
We reverted the PR with native-stack because it had too many issues - https://github.com/Expensify/App/pull/37305

I'll create a new PR where I'll resolve all issues 👍

**Comment by mountiny** [comment]
Thanks!

**Comment by abdulrahuman5196** [comment]
Not awaiting payment. Still in PR phase since the PR got reverted.

**Comment by mountiny** [comment]
@kirillzyusko whats your ETA for this one?

**Comment by kirillzyusko** [comment]
@mountiny I think early next week (Monday, Tuesday) my PR will be ready for review and testing. I fixed all the issues that were reported for previous PR, but I tested apps side-by-side (old stack navigator and native-stack navigator) and discovered more issues which I'm fixing right now - but the progress is moving pretty good, so I think I'll be able to make something in a good shape next Monday/Tuesday (but we'll see how it goes).

**Comment by mountiny** [comment]
We should be merging the Fabric branch this week

**Comment by quinthar** [comment]
Should this be closed?  If not, what is the next step, who is doing it, and when?

**Comment by mountiny** [comment]
The PR is actively being worked on https://github.com/Expensify/App/pull/37891, we are addressing some changes required upstream

**Comment by kirillzyusko** [comment]
@mountiny should we re-open this?

**Comment by mountiny** [comment]
Working on the PR again

**Comment by mountiny** [comment]
Not awaiting for payment yet, we split this up to two PRs

**Comment by anmurali** [comment]
@mountiny is this still just on hold? I am not supposed to be paying anyone?

**Comment by anmurali** [comment]
Rotating to get another BZ assigned while OOO

**Comment by mountiny** [comment]
Not supposed to pay yet, updated the title

**Comment by joekaufmanexpensify** [comment]
Sounds good!

**Comment by mountiny** [comment]
Not on hold, we are working through it

**Comment by joekaufmanexpensify** [comment]
Still in progress

**Comment by mountiny** [comment]
This was merged and now its in production 🎉

**Comment by joekaufmanexpensify** [comment]
Going to issue payment here in the next day!

**Comment by mountiny** [comment]
PRs to pay out:
- https://github.com/Expensify/App/pull/29991 - there was many regressions and it was reverted I think no payments required there
- https://github.com/Expensify/App/pull/37891 - @ishpaul777 
- https://github.com/Expensify/App/pull/49937 - @ishpaul777 and @c3024 

I would say $1000 to @ishpaul777 and $250 to @c3024 for their work on these

**Comment by joekaufmanexpensify** [comment]
Got it. TY! Opening new upwork job so we can pay.

**Comment by joekaufmanexpensify** [comment]
Ah, actually noticed @c3024 is already being paid for their PR [here](https://github.com/Expensify/App/issues/53264). So we just need to pay @ishpaul777

**Comment by joekaufmanexpensify** [comment]
@ishpaul777 offer sent in Upwork!

**Comment by mountiny** [comment]
@ishpaul777 ah nice, thanks for the link. Let's increase to $1250, including that PR. Thanks!

**Comment by joekaufmanexpensify** [comment]
Sounds good! Adjusted payment summary. @ishpaul777 please accept the offer I sent. I will pay the extra $250 as a bonus.

**Comment by joekaufmanexpensify** [comment]
@ishpaul777 $1,250 sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
All set, thanks everyone!
